sphinx/util/inspect.py

```diff
@@ -125,13 +125,15 @@ def unwrap(obj: Any) -> Any:
         return obj
 
 
-def unwrap_all(obj: Any) -> Any:
+def unwrap_all(obj: Any, *, stop: Callable = None) -> Any:
     """
     Get an original object from wrapped object (unwrapping partials, wrapped
     functions, and other decorators).
     """
     while True:
-        if ispartial(obj):
+        if stop and stop(obj):
+            return obj
+        elif ispartial(obj):
```

Comment:
Can you add a test of a partial async function?

Reply:
We already have it 
https://github.com/sphinx-doc/sphinx/blob/017af4ce2ef1fbc4387100b850e202b17f4c7450/tests/test_util_inspect.py#L473